rasa/telemetry.py

```diff
@@ -0,0 +1,553 @@
+TRAINING_COMPLETED_EVENT = "Training Completed"
+TELEMETRY_DISABLED = "Telemetry Disabled"
+
+
+def print_telemetry_reporting_info() -> None:
+    """Print telemetry information to std out."""
+    message = textwrap.dedent(
+        f"""
+      Rasa reports anonymous usage telemetry to help improve Rasa Open Source
+      for all its users.
+
+      If you'd like to opt-out, you can use `rasa telemetry disable`.
+      To learn more, check out {DOCS_URL_TELEMETRY}."""
+    ).strip()
+
+    table = SingleTable([[message]])
+    print(table.table)
+
+
+def _default_telemetry_configuration(is_enabled: bool) -> Dict[Text, Text]:
+    return {
+        CONFIG_TELEMETRY_ENABLED: is_enabled,
+        CONFIG_TELEMETRY_ID: uuid.uuid4().hex,
+        CONFIG_TELEMETRY_DATE: datetime.now(),
+    }
+
+
+def _write_default_telemetry_configuration(
+    is_enabled: bool = TELEMETRY_ENABLED_BY_DEFAULT,
+) -> None:
+    if is_enabled:
+        print_telemetry_reporting_info()
+
+    new_config = _default_telemetry_configuration(is_enabled)
+
+    rasa_utils.write_global_config_value(CONFIG_FILE_TELEMETRY_KEY, new_config)
+
+
+def _is_telemetry_enabled_in_configuration() -> bool:
+    """Read telemetry configuration from the user's Rasa config file in $HOME.
+
+    Creates a default configuration if no configuration exists.
+
+    Returns:
+        `True`, if telemetry is enabled, `False` otherwise.
+    """
+    try:
+        stored_config = rasa_utils.read_global_config_value(
+            CONFIG_FILE_TELEMETRY_KEY, unavailable_ok=False
+        )
+
+        return stored_config[CONFIG_TELEMETRY_ENABLED]
+    except ValueError as e:
+        logger.debug(f"Could not read telemetry settings from configuration file: {e}")
+
+        # seems like there is no config, we'll create on and enable telemetry
+        _write_default_telemetry_configuration()
+        return TELEMETRY_ENABLED_BY_DEFAULT
+
+
+def initialize_telemetry() -> bool:
+    """Read telemetry configuration from the user's Rasa config file in $HOME.
+
+    Creates a default configuration if no configuration exists.
+
+    Returns:
+        `True`, if telemetry is enabled, `False` otherwise.
+    """
+
+    # calling this even if the environment variable is set makes sure the
+    # configuration is created and there is a telemetry ID
+    is_enabled_in_configuration = _is_telemetry_enabled_in_configuration()
+
+    telemetry_environ = os.environ.get(TELEMETRY_ENABLED_ENVIRONMENT_VARIABLE)
+
+    if telemetry_environ is None:
+        return is_enabled_in_configuration
+    else:
+        return telemetry_environ.lower() == "true"
+
+
+def ensure_telemetry_enabled(f: Callable[..., Any]) -> Callable[..., Any]:
+    """Function decorator for telemetry functions that only runs the decorated
+    function if telemetry is enabled."""
+
+    is_telemetry_enabled = initialize_telemetry()
+
+    # allows us to use the decorator for async and non async functions
+    if asyncio.iscoroutinefunction(f):
```

Comment:
we could also use `asyncio.coroutine` to wrap any non async functions which in turn would mean that we could deduplicate the block below. Or alternatively moving the `asyncio.iscoroutinefunction` within the decorator

Reply:
`asyncio.coroutine` is deprecated

Reply:
then let's use an `if ... else` in the decorator?

Reply:
@tmbo

Reply:
I am not quite sure what you mean 🤔 `ensure_telemetry_enabled ` is the decorator. do you mean inside of `decorated`? I don't think that would work since that function needs to be either `async def` or `def`

Reply:
Never mind :-) Played a bit around with it, and it's surprisingly complicated / impossible to deduplicate it further for this particular use case.